File: backend/src/utils/response.ts
// utils/response.ts
import { Response } from 'express';

export class ApiResponse {
  // Support both signatures:
  // ApiResponse.success(res, data, message?)
  // ApiResponse.success(res, message, data?)
  static success(res: Response, a: any, b: any = undefined, statusCode = 200) {
    let data: any;
    let message: string;

    // If second arg is a string, treat as message: (res, message, data?)
    if (typeof a === 'string') {
      message = a;
      data = b;
    } else {
      // Otherwise treat as (res, data, message?)
      data = a;
      message = typeof b === 'string' ? b : 'Success';
    }

    return res.status(statusCode).json({
      success: true,
      message,
      data
    });
  }

  static error(res: Response, message = 'Error', statusCode = 500, errors?: any) {
    return res.status(statusCode).json({
      success: false,
      message,
      errors
    });
  }

  static paginated(res: Response, data: any[], pagination: any, message = 'Success') {
    return res.status(200).json({
      success: true,
      message,
      data,
      pagination
    });
  }
}

File: backend-rust/src/utils/response.rs
use axum::{
    http::StatusCode,
    response::{IntoResponse, Response},
    Json,
};
use serde::Serialize;

#[derive(Serialize)]
pub struct ApiResponse<T> {
    pub success: bool,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<T>,
}

#[derive(Serialize)]
pub struct PaginationMetadata {
    pub page: u32,
    pub limit: u32,
    pub total: u64,
    pub pages: u32,
}

#[derive(Serialize)]
pub struct PaginatedResponse<T> {
    pub success: bool,
    pub message: String,
    pub data: Vec<T>,
    pub pagination: PaginationMetadata,
}

impl ApiResponse<()> {
    pub fn error(message: &str, status: StatusCode) -> Response {
        (
            status,
            Json(ApiResponse::<()> {
                success: false,
                message: message.to_string(),
                data: None,
            }),
        )
            .into_response()
    }
}

impl<T: Serialize> ApiResponse<T> {
    pub fn success(data: T, message: &str, status: StatusCode) -> Response {
        (
            status,
            Json(ApiResponse {
                success: true,
                message: message.to_string(),
                data: Some(data),
            }),
        )
            .into_response()
    }

    pub fn paginated(
        data: Vec<T>,
        meta: PaginationMetadata,
        message: &str,
        status: StatusCode,
    ) -> Response {
        (
            status,
            Json(PaginatedResponse {
                success: true,
                message: message.to_string(),
                data,
                pagination: meta,
            }),
        )
            .into_response()
    }
}
